### ISSUE TITLE
Google session lifetime is too short

### ISSUE THREAD

**Issue by dsavenko** [remotestorage/remotestorage.js#1159]
Can I configure Google session lifetime? It's a couple of hours or so. After that the user's automatically asked to re-login. I'd prefer the session to last nearly forever, as it is for the Google Drive website itself.

**Comment by skddc** [comment]
Not an option in this library at the moment, but if you can find out how to do it with their API, that could be a great improvement.

**Comment by dsavenko** [comment]
Found this SO thread: https://stackoverflow.com/a/13851781
Is it helpful?

**Comment by iLiviu** [comment]
Renewing access tokens would require exposing the OAuth client secret in your app. This is highly not recommended in client-side browser apps.

**Comment by skddc** [comment]
Looks like this is a shortcoming of the Google OAuth 2.0 API for web apps. If an app is configured as Android, iOS or Chrome app, then it doesn't need the secret for acquiring refresh tokens, according to https://developers.google.com/identity/protocols/OAuth2InstalledApp

Afaics it's also not possible to configure the token expiry time. This leaves us with no options here, until Google decides to loosen the rules for web apps (hopefully at least for installed PWAs).

**Comment by dsavenko** [comment]
I don't think this is a problem. According to [Google docs](https://developers.google.com/api-client-library/javascript/start/start-js), you need API key and client ID, both can be exposed (as long as you configured them to accept only certain domains and referrers). 

On the other hand, making users re-login every hour makes Google integration almost unusable.

**Comment by iLiviu** [comment]
Not being able to auto-renew access tokens for client side JavaScript apps is a feature of OAuth, not a bug. Access tokens should be temporary and valid only when user is present in front of the device, that is why they expire after a few hours. The user should give his consent each time he logs in again.  

Access tokens can only be renewed automatically (without user consent), by using the client secret, which should only be used by server-side apps, that can keep the client secret safe.

**Comment by skddc** [comment]
> Access tokens can only be renewed automatically (without user consent), by using the client secret, which should only be used by server-side apps, that can keep the client secret safe.

Yes, but that's why Google (and others) have an exception for not needing a secret in native mobile apps. Those are also client-side, so it would possible to fish out the secret otherwise. When the standard was written, nobody thought of it being used with all these client-side apps, or that PWAs would even be a thing eventually.

The problem, afaics, is that according to their docs, they don't allow the same exception for PWAs, that they do for native mobile apps.

@dsavenko The doc you linked doesn't say anything about refreshing tokens. But if you're sure it's possible without a *secret* (not key), then maybe you could link to code that does it?

**Comment by iLiviu** [comment]
> The problem, afaics, is that according to their docs, they don't allow the same exception for PWAs, that they do for native mobile apps. (Please correct me if that's wrong.)

As i see it, client side Javascript applications would never need refresh tokens, since once the the auth token expires, they can just start the authorization process again, and if the user is already logged into his google account in the browser, the authorization process will not require any user interaction. I checked now and the GoogleDrive backend already does this (it detects when token is expired and starts authorization process again), the only inconvenience is that the page is refreshed when doing so.

**Comment by skddc** [comment]
Yes, it's definitely all about that page refresh. We implemented the re-auth last year, so that's intentionally working as it is. Personally, I just didn't know how short-lived the tokens are. So the refresh basically happens every time you open an app, which is definitely not ideal UX.

**Comment by iLiviu** [comment]
A solution would be the auth to be done in a new window that automatically closes when it's done.  I noticed someone else already opened an issue about this.

**Comment by skddc** [comment]
That would still leave the bad UX of opening and immediately closing a window every time you open the app.

**Comment by iLiviu** [comment]
It could be done in a hidden IFRAME. And the good news is that Google (or better said 
OpenID Connect protocol which Google implements) allows to pass a `prompt` parameter  in the authorization request, and set it to `none`, which instructs Google to not display any authentication or consent screens during auth. This means that if the user has not authorized the app yet or is not logged into his google account, it will redirect with an error message immediately. The library can pick the error and proceed with normal authorization in the parent window. Otherwise, if the user is logged into Google and already authorized the app, a new token will be granted and parent window can pick it up without user being affected at all.

**Comment by skddc** [comment]
I was rather sure you cannot do that with an iFrame, because iFrames are intentionally sandboxed from the parent window. Do you have a code example of this actually working?

**Comment by iLiviu** [comment]
IFRAMEs are not sandboxed if parent is from same domain. However, it's possible to communicate even if the IFRAME location is from different domain (although not the case here), using [window.PostMessage()](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) (there is an example of two-way communication there).

**Comment by skddc** [comment]
OK, cool. So if it's possible to redirect to Google in a hidden iFrame, then this could be a solution indeed. It's much more hacky than just refreshing the token via the normal API, like native mobile apps do, but at least we could get rid of the redirect this way.
